### PR TITLE
fix dirty data in peerset,resolve #3304

### LINKED ISSUE #3304: PeerSet contains dirty peer data cause reconnection failed
**Description:** 
The topology is:
A seed node connected to serval validator nodes(Exactly two nodes). 
Two validator nodes is persistent node of seed node.
After running for serval minutes, seed nodes lose all connection to validator, and try to reconnect,
but all failed, and seed node consensus no blocks any more.
  

## Steps To Reproduce:

  1. Deploy a validator network.
  2. Deploy a seed node(seed_mode is true), connected to core network.
  3. Wait for some time or just use `iptable` to produce network joggle.

## Supporting Material/References:
We check the validator logs:
```
I[2019-02-01|13:32:37.737] Starting Peer                                module=p2p peer=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e@104.193.88.123:27146 impl="Peer{MConn{10.201.41.209:2494} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in}"
I[2019-02-01|13:32:37.737] Starting MConnection                         module=p2p peer=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e@104.193.88.123:27146 impl=MConn{10.201.41.209:2494}
I[2019-02-01|13:32:37.739] Added peer                                   module=p2p peer="Peer{MConn{10.201.41.209:2494} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in}"
E[2019-02-01|13:32:38.125] Stopping peer for error                      module=p2p peer="Peer{MConn{10.201.41.209:2494} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in}" err="Peer (77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e) sent next PEX request too soon. lastReceived: 2019-02-01 13:32:38.016135243 +0000 UTC m=+41.575722596, now: 2019-02-01 13:32:38.125079205 +0000 UTC m=+41.684666557, minInterval: 10s. Disconnecting"
I[2019-02-01|13:32:38.125] Stopping Peer                                module=p2p peer=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e@104.193.88.123:27146 impl="Peer{MConn{10.201.41.209:2494} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in}"
I[2019-02-01|13:32:38.125] Stopping MConnection                         module=p2p peer=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e@104.193.88.123:27146 impl=MConn{10.201.41.209:2494}
I[2019-02-01|13:32:38.133] Stopping gossipDataRoutine for peer          module=consensus peer="Peer{MConn{10.201.41.209:2494} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in}"
I[2019-02-01|13:32:38.134] Stopping gossipVotesRoutine for peer         module=consensus peer="Peer{MConn{10.201.41.209:2494} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in}"
I[2019-02-01|13:32:38.930] Starting Peer                                module=p2p peer=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e@119.63.197.139:27146 impl="Peer{MConn{10.201.43.54:2695} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in}"
I[2019-02-01|13:32:38.930] Starting MConnection                         module=p2p peer=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e@119.63.197.139:27146 impl=MConn{10.201.43.54:2695}
I[2019-02-01|13:32:39.157] Added peer                                   module=p2p peer="Peer{MConn{10.201.43.54:2695} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in}"
I[2019-02-01|13:32:39.159] Ignoring inbound connection: error while adding peer module=p2p err="duplicate ID<77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e>" id=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e
I[2019-02-01|13:32:39.737] Stopping queryMaj23Routine for peer          module=consensus peer="Peer{MConn{10.201.41.209:2494} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in}"
I[2019-02-01|13:33:06.788] Dialing peer                                 module=p2p address=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e@52.196.133.127:27146
E[2019-02-01|13:33:06.789] Dialing failed                               module=pex addr=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e@52.196.133.127:27146 err="auth failure: secrect conn failed: EOF" attempts=0
I[2019-02-01|13:33:09.127] Starting Peer                                module=p2p peer=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e@119.63.197.151:27146 impl="Peer{MConn{10.201.42.180:2682} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in}"
I[2019-02-01|13:33:09.127] Starting MConnection                         module=p2p peer=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e@119.63.197.151:27146 impl=MConn{10.201.42.180:2682}
I[2019-02-01|13:33:09.227] Stopping MConnection                         module=p2p peer=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e@119.63.197.151:27146 impl=MConn{10.201.42.180:2682}
E[2019-02-01|13:33:09.227] Stopping peer for error                      module=p2p peer="Peer{MConn{10.201.42.180:2682} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in}" err="Peer Peer{MConn{10.201.42.180:2682} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in} has no state"
I[2019-02-01|13:33:09.227] Stopping Peer                                module=p2p peer=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e@119.63.197.151:27146 impl="Peer{MConn{10.201.42.180:2682} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in}"
I[2019-02-01|13:33:09.402] Added peer                                   module=p2p peer="Peer{MConn{10.201.42.180:2682} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in}"
I[2019-02-01|13:33:09.403] Stopping queryMaj23Routine for peer          module=consensus peer="Peer{MConn{10.201.42.180:2682} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in}"
I[2019-02-01|13:33:09.403] Stopping gossipVotesRoutine for peer         module=consensus peer="Peer{MConn{10.201.42.180:2682} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in}"
I[2019-02-01|13:33:09.403] Stopping gossipDataRoutine for peer          module=consensus peer="Peer{MConn{10.201.42.180:2682} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in}"
I[2019-02-01|13:33:09.570] Ignoring inbound connection: error while adding peer module=p2p err="duplicate ID<77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e>" id=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e
I[2019-02-01|13:33:09.836] Ignoring inbound connection: error while adding peer module=p2p err="duplicate ID<77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e>" id=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e
I[2019-02-01|13:33:10.103] Ignoring inbound connection: error while adding peer module=p2p err="duplicate ID<77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e>" id=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e
I[2019-02-01|13:33:10.367] Ignoring inbound connection: error while adding peer module=p2p err="duplicate ID<77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e>" id=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e
I[2019-02-01|13:33:10.639] Ignoring inbound connection: error while adding peer module=p2p err="duplicate ID<77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e>" id=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e
I[2019-02-01|13:33:11.178] Ignoring inbound connection: error while adding peer module=p2p err="duplicate ID<77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e>" id=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e
I[2019-02-01|13:33:11.448] Ignoring inbound connection: error while adding peer module=p2p err="duplicate ID<77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e>" id=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e
I[2019-02-01|13:33:11.719] Ignoring inbound connection: error while adding peer module=p2p err="duplicate ID<77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e>" id=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e
I[2019-02-01|13:33:11.982] Ignoring inbound connection: error while adding peer module=p2p err="duplicate ID<77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e>" id=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e
I[2019-02-01|13:33:12.245] Ignoring inbound connection: error while adding peer module=p2p err="duplicate ID<77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e>" id=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e
I[2019-02-01|13:33:12.288] Ignoring inbound connection: error while adding peer module=p2p err="duplicate ID<77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e>" id=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e
I[2019-02-01|13:33:12.509] Ignoring inbound connection: error while adding peer module=p2p err="duplicate ID<77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e>" id=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e
I[2019-02-01|13:33:12.784] Ignoring inbound connection: error while adding peer module=p2p err="duplicate ID<77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e>" id=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e
I[2019-02-01|13:33:13.046] Ignoring inbound connection: error while adding peer module=p2p err="duplicate ID<77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e>" id=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e
```

We can see the incoming connection is disconnected by validator since it report `duplicate ID`.
We can find that the sequence is not correct, `Added peer` is after `Stopping Peer`: 
```
I[2019-02-01|13:33:09.127] Starting MConnection                         module=p2p peer=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e@119.63.197.151:27146 impl=MConn{10.201.42.180:2682}
I[2019-02-01|13:33:09.227] Stopping MConnection                         module=p2p peer=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e@119.63.197.151:27146 impl=MConn{10.201.42.180:2682}
E[2019-02-01|13:33:09.227] Stopping peer for error                      module=p2p peer="Peer{MConn{10.201.42.180:2682} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in}" err="Peer Peer{MConn{10.201.42.180:2682} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in} has no state"
I[2019-02-01|13:33:09.227] Stopping Peer                                module=p2p peer=77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e@119.63.197.151:27146 impl="Peer{MConn{10.201.42.180:2682} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in}"
I[2019-02-01|13:33:09.402] Added peer                                   module=p2p peer="Peer{MConn{10.201.42.180:2682} 77d3f2ee9b5edfd32ffc017e7e8f9e4a1dc1457e in}"
```

Refer to the code, `sw.peers.Add(p)` happens after `sw.startInitPeer(p)`:
```
func (sw *Switch) addPeer(p Peer) error {
	if err := sw.filterPeer(p); err != nil {
		return err
	}

	p.SetLogger(sw.Logger.With("peer", p.NodeInfo().NetAddress()))

	// Handle the shut down case where the switch has stopped but we're
	// concurrently trying to add a peer.
	if sw.IsRunning() {
		// All good. Start peer
		if err := sw.startInitPeer(p); err != nil {
			return err
		}
	} else {
		sw.Logger.Error("Won't start a peer - switch is not running", "peer", p)
	}

	// Add the peer to .peers.
	// We start it first so that a peer in the Alist is safe to Stop.
	// It should not err since we already checked peers.Has().
	if err := sw.peers.Add(p); err != nil {
		return err
	}

	sw.Logger.Info("Added peer", "peer", p)
	sw.metrics.Peers.Add(float64(1))
	return nil
```
We know `sw.startInitPeer(p)` will enable every reactor to send/receive message with peer.
In the log our peer get wrong message and disconnect the peer by callling `StopPeerForError`:
```
func(sw*Switch)StopPeerForError(peerPeer,reasoninterface{}){
sw.Logger.Error("Stoppingpeerforerror","peer",peer,"err",reason,"IsPersistent",peer.IsPersistent())
sw.stopAndRemovePeer(peer,reason)

ifpeer.IsPersistent(){
//FIXME:persistentpeerscan'tbeinboundrightnow.
//self-reportedaddressforinboundpersistentpeers
addr:=peer.NodeInfo().NetAddress()
gosw.reconnectToPeer(addr)
}
}
```
Which will 
```
func(ps*PeerSet)Remove(peerPeer)bool{
ps.mtx.Lock()
deferps.mtx.Unlock()
//fmt.Printf("===debugpeersetrmid%v\n",peer.ID())
item:=ps.lookup[peer.ID()]
ifitem==nil{
returnfalse
}
```
But the peer is still not added to `PeerSet`, so it remove nothing.

Then in the `func (sw *Switch) addPeer(p Peer) ` will add the peer to peerset.
Nexttime the peer try to connect to our peer, it will be rejected. That is why we see `duplicate ID` error

In the log, we see `no state error`, and the mconnection goroutine panic.
The reason why get "no state" error:
The state is set:
```
func (conR *ConsensusReactor) AddPeer(peer p2p.Peer) {
	if !conR.IsRunning() {
		return
	}

	// Create peerState for peer
	peerState := NewPeerState(peer).SetLogger(conR.Logger)
	peer.Set(types.PeerStateKey, peerState)

	// Begin routines for this peer.
	go conR.gossipDataRoutine(peer, peerState)
	go conR.gossipVotesRoutine(peer, peerState)
	go conR.queryMaj23Routine(peer, peerState)

	// Send our state to peer.
	// If we're fast_syncing, broadcast a RoundStepMessage later upon SwitchToConsensus().
	if !conR.FastSync() {
		conR.sendNewRoundStepMessage(peer)
	}
}
```
but mconnection is starting before consensus reactor staring:
```
func (sw *Switch) startInitPeer(p Peer) error {
	err := p.Start() // spawn send/recv routines
	if err != nil {
		// Should never happen
		sw.Logger.Error(
			"Error starting peer",
			"err", err,
			"peer", p,
		)
		return err
	}

	for _, reactor := range sw.reactors {
		reactor.AddPeer(p)
	}

	return nil
}
```
Onec mconnection get consensus message and AddPeer is not completed, error happened

### PR DESCRIPTION
**Pull request by guagualvcha**
Fix reconnection report duplicate id error.

startInitPeer before PeerSet add the peer,
once mconnection start and Receive of one
Reactor faild, will try to remove it from PeerSet
while PeerSet still not contain the peer. Fix
this by change the order.

* [x] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [x] Wrote tests
* [x] Updated CHANGELOG_PENDING.md

### CODE REVIEW

**Comment by guagualvcha** [comment]
how can i rerun the ci?

**Comment by ebuchman** [comment]
Looks like there's a data race:

```
==================
WARNING: DATA RACE
Read at 0x00c000140508 by goroutine 124:
  github.com/tendermint/tendermint/p2p/conn.(*MConnection).stopServices()
      /go/src/github.com/tendermint/tendermint/p2p/conn/connection.go:222 +0x93
  github.com/tendermint/tendermint/p2p/conn.(*MConnection).OnStop()
      /go/src/github.com/tendermint/tendermint/p2p/conn/connection.go:278 +0x38
  github.com/tendermint/tendermint/libs/common.(*BaseService).Stop()
      /go/src/github.com/tendermint/tendermint/libs/common/service.go:167 +0x4be
  github.com/tendermint/tendermint/p2p.(*peer).OnStop()
      /go/src/github.com/tendermint/tendermint/p2p/peer.go:200 +0x7c
  github.com/tendermint/tendermint/libs/common.(*BaseService).Stop()
      /go/src/github.com/tendermint/tendermint/libs/common/service.go:167 +0x4be
  github.com/tendermint/tendermint/p2p.(*peer).Stop()
      <autogenerated>:1 +0x43
D[2019-02-28|09:56:38.719] Stopping MConnection (ignoring: already stopped) switch=2 peer=4c770204e88a075cf92e5fe9c0a9bcad8fc8e3dd@127.0.0.1:43486 impl=MConn{127.0.0.1:37046}
  github.com/tendermint/tendermint/p2p.(*Switch).OnStop()
      /go/src/github.com/tendermint/tendermint/p2p/switch.go:214 +0x129
  github.com/tendermint/tendermint/libs/common.(*BaseService).Stop()
      /go/src/github.com/tendermint/tendermint/libs/common/service.go:167 +0x4be
  github.com/tendermint/tendermint/p2p/pex.TestConnectionSpeedForPeerReceivedFromSeed()
      /go/src/github.com/tendermint/tendermint/p2p/pex/pex_reactor_test.go:290 +0x4d4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162

Previous write at 0x00c000140508 by goroutine 108:
  github.com/tendermint/tendermint/p2p/conn.(*MConnection).OnStart()
      /go/src/github.com/tendermint/tendermint/p2p/conn/connection.go:207 +0x253
  github.com/tendermint/tendermint/libs/common.(*BaseService).Start()
      /go/src/github.com/tendermint/tendermint/libs/common/service.go:139 +0x4df
  github.com/tendermint/tendermint/p2p.(*peer).OnStart()
      /go/src/github.com/tendermint/tendermint/p2p/peer.go:179 +0x56
  github.com/tendermint/tendermint/libs/common.(*BaseService).Start()
      /go/src/github.com/tendermint/tendermint/libs/common/service.go:139 +0x4df
  github.com/tendermint/tendermint/p2p.(*peer).Start()
      <autogenerated>:1 +0x43
  github.com/tendermint/tendermint/p2p.(*Switch).startInitPeer()
      /go/src/github.com/tendermint/tendermint/p2p/switch.go:671 +0x5c
  github.com/tendermint/tendermint/p2p.(*Switch).addPeer()
      /go/src/github.com/tendermint/tendermint/p2p/switch.go:656 +0x55d
  github.com/tendermint/tendermint/p2p.(*Switch).addOutboundPeerWithConfig()
      /go/src/github.com/tendermint/tendermint/p2p/switch.go:598 +0x702
  github.com/tendermint/tendermint/p2p.(*Switch).DialPeerWithAddress()
      /go/src/github.com/tendermint/tendermint/p2p/switch.go:457 +0x144
E[2019-02-28|09:56:38.719] Connection failed @ recvRoutine (reading byte) switch=1 peer=4c770204e88a075cf92e5fe9c0a9bcad8fc8e3dd@127.0.0.1:43486 conn=MConn{127.0.0.1:51244} err=EOF
  github.com/tendermint/tendermint/p2p/pex.(*PEXReactor).dialPeer()
      /go/src/github.com/tendermint/tendermint/p2p/pex/pex_reactor.go:494 +0x13c

Goroutine 124 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:878 +0x659
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1119 +0xa8
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1117 +0x4ee
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1034 +0x2ee
  main.main()
      _testmain.go:156 +0x332

Goroutine 108 (running) created at:
  github.com/tendermint/tendermint/p2p/pex.(*PEXReactor).ensurePeers()
      /go/src/github.com/tendermint/tendermint/p2p/pex/pex_reactor.go:440 +0x992
  github.com/tendermint/tendermint/p2p/pex.(*PEXReactor).ReceiveAddrs()
      /go/src/github.com/tendermint/tendermint/p2p/pex/pex_reactor.go:341 +0x417
  github.com/tendermint/tendermint/p2p/pex.(*PEXReactor).Receive()
      /go/src/github.com/tendermint/tendermint/p2p/pex/pex_reactor.go:248 +0xebe
  github.com/tendermint/tendermint/p2p.createMConnection.func1()
      /go/src/github.com/tendermint/tendermint/p2p/peer.go:374 +0x25b
  github.com/tendermint/tendermint/p2p/conn.(*MConnection).recvRoutine()
      /go/src/github.com/tendermint/tendermint/p2p/conn/connection.go:599 +0xcce
==================
```

Also some consensus failure logs but not sure if those are real problem or artifacts of test:

```
=== RUN   TestProvider
--- PASS: TestProvider (5.59s)
    provider_test.go:37: chainID: tendermint_test
PASS
coverage: 56.5% of statements
E[2019-02-28|09:56:10.399] CONSENSUS FAILURE!!!                         module=consensus err="open /tmp/-_go_src_github.com_tendermint_tendermint_lite_client_836874544/data/write-file-atomic-6169177413388503205: no such file or directory" stack="goroutine 88 [running]:\nruntime/debug.Stack(0xc001100e08, 0x10b3580, 0xc0013409f0)\n\t/usr/local/go/src/runtime/debug/stack.go:24 +0xb5\ngithub.com/tendermint/tendermint/consensus.(*ConsensusState).receiveRoutine.func2(0xc0009a8e00, 0x1308eb0)\n\t/go/src/github.com/tendermint/tendermint/consensus/state.go:611 +0x6a\npanic(0x10b3580, 0xc0013409f0)\n\t/usr/local/go/src/runtime/panic.go:513 +0x1b9\ngithub.com/tendermint/tendermint/privval.(*FilePVLastSignState).Save(0xc00098aea8)\n\t/go/src/github.com/tendermint/tendermint/privval/file.go:126 +0x1a5\ngithub.com/tendermint/tendermint/privval.(*FilePV).saveSigned(0xc00098ae60, 0xc6, 0x0, 0x2, 0xc000f59400, 0x75, 0xf4, 0xc000b75700, 0x40, 0x40)\n\t/go/src/github.com/tendermint/tendermint/privval/file.go:372 +0x10f\ngithub.com/tendermint/tendermint/privval.(*FilePV).signVote(0xc00098ae60, 0xc0009ffc60, 0xf, 0xc000b34320, 0x16911ac9, 0xed409a73a)\n\t/go/src/github.com/tendermint/tendermint/privval/file.go:316 +0x5c8\ngithub.com/tendermint/tendermint/privval.(*FilePV).SignVote(0xc00098ae60, 0xc0009ffc60, 0xf, 0xc000b34320, 0x0, 0xc000c76cc0)\n\t/go/src/github.com/tendermint/tendermint/privval/file.go:239 +0x64\ngithub.com/tendermint/tendermint/consensus.(*ConsensusState).signVote(0xc0009a8e00, 0xc00133c401, 0xc00133c4a0, 0x20, 0x20, 0x1, 0xc0013644e0, 0x20, 0x20, 0xc00097c5a0, ...)\n\t/go/src/github.com/tendermint/tendermint/consensus/state.go:1695 +0x44e\ngithub.com/tendermint/tendermint/consensus.(*ConsensusState).signAddVote(0xc0009a8e00, 0xc00133c401, 0xc00133c4a0, 0x20, 0x20, 0x1, 0xc0013644e0, 0x20, 0x20, 0xc0011c9640)\n\t/go/src/github.com/tendermint/tendermint/consensus/state.go:1722 +0x1c6\ngithub.com/tendermint/tendermint/consensus.(*ConsensusState).defaultDoPrevote(0xc0009a8e00, 0xc6, 0x0)\n\t/go/src/github.com/tendermint/tendermint/consensus/state.go:1030 +0x77e\ngithub.com/tendermint/tendermint/consensus.(*ConsensusState).defaultDoPrevote-fm(0xc6, 0x0)\n\t/go/src/github.com/tendermint/tendermint/consensus/state.go:169 +0x56\ngithub.com/tendermint/tendermint/consensus.(*ConsensusState).enterPrevote(0xc0009a8e00, 0xc6, 0x0)\n\t/go/src/github.com/tendermint/tendermint/consensus/state.go:994 +0x583\ngithub.com/tendermint/tendermint/consensus.(*ConsensusState).addProposalBlockPart(0xc0009a8e00, 0xc00000dc00, 0x0, 0x0, 0xc0004fd380, 0xc000a86078, 0xc001101a00)\n\t/go/src/github.com/tendermint/tendermint/consensus/state.go:1491 +0xaf6\ngithub.com/tendermint/tendermint/consensus.(*ConsensusState).handleMsg(0xc0009a8e00, 0x1408a80, 0xc00000dc00, 0x0, 0x0)\n\t/go/src/github.com/tendermint/tendermint/consensus/state.go:682 +0xb04\ngithub.com/tendermint/tendermint/consensus.(*ConsensusState).receiveRoutine(0xc0009a8e00, 0x0)\n\t/go/src/github.com/tendermint/tendermint/consensus/state.go:655 +0x6ae\ncreated by github.com/tendermint/tendermint/consensus.(*ConsensusState).OnStart\n\t/go/src/github.com/tendermint/tendermint/consensus/state.go:334 +0x3b1\n"
E[2019-02-28|09:56:10.399] Not stopping BlockPool -- have not been started yet module=blockchain impl=BlockPool
E[2019-02-28|09:56:10.399] Stopped accept routine, as transport is closed module=p2p numPeers=0
E[2019-02-28|09:56:10.400] Failed to save AddrBook to file              module=p2p book=/tmp/-_go_src_github.com_tendermint_tendermint_lite_client_836874544/config/addrbook.json file=/tmp/-_go_src_github.com_tendermint_tendermint_lite_client_836874544/config/addrbook.json err="open /tmp/-_go_src_github.com_tendermint_tendermint_lite_client_836874544/config/write-file-atomic-04407214388515413680: no such file or directory"
ok  	github.com/tendermint/tendermint/lite/client	7.154s	coverage: 56.5% of statements
=== RUN   TestNodeStartStop
```

I'll re run it

**Comment by ebuchman** [comment]
It looks like connection.Start() and connection.Stop() are being called concurrently, which is a violation of the BaseService contract which states: `The caller must ensure that Start and Stop are not called concurrently.` With the code written as before, we started the peer before adding it to the set, so we knew we wouldnt end up trying to stop it concurrently with start. Now that we add it first, Stop can be called on it. Will need to think more.

**Comment by ebuchman** [comment]
Oh I think we just need to break up startInitPeer - so the order of events would be:
1. `p.Start()`
2. `peers.Add(p)`
3. `reactor.AddPeer(p)`

File: p2p/switch.go [hunk 1]
```diff
@@ -646,21 +646,21 @@ func (sw *Switch) addPeer(p Peer) error {
 	// Handle the shut down case where the switch has stopped but we're
 	// concurrently trying to add a peer.
 	if sw.IsRunning() {
-		// All good. Start peer
+		// Add the peer to PeerSet.
+		// We add peer first, if start peer failed, need remove it.
```
Comment (melekes) [inline p2p/switch.go]: Would be good to explain why, not spell what the code does.

**Comment by ebuchman** [comment]
Ok I think this is fixed now.

File: p2p/switch_test.go [hunk 2]
```diff
@@ -273,6 +273,8 @@ func TestSwitchPeerFilterTimeout(t *testing.T) {
 
 func TestSwitchPeerFilterDuplicate(t *testing.T) {
 	sw := MakeSwitch(cfg, 1, "testing", "123.123.123", initSwitchFunc)
+	sw.Start()
```
Comment (melekes) [inline p2p/switch_test.go]:
```suggestion
	err := sw.Start()
	require.NoError(t, err)
```

**Comment by guagualvcha** [comment]
@melekes  @ebuchman   I think the following commit can't  add peer to peerset after peer start.  The main purpose of this pr is to add peer to peerset before peer start to avod dirty data in peerset.

**Comment by ebuchman** [comment]
I see. I suppose the original issue (the peer errs before it even gets added to the peer set) is still possible, but should now be much less likely since we don't start the peer on the reactors until after its added to the peer set.

We'll have to think more on a proper solution as there are a bunch of concurrency issues here. The original solution was causing test failures as mentioned due to Start/Stop being possibly called concurrently.

**Pull request merged**